File: src/Algorithms/Lists/Rotation_Point/rotation-point.ts
// O (log(n)) time
// O(1) space
const solution = (words: string[]) => {
  const firstWord = words[0];

  let floorIndex = 0;
  let ceilIndex = words.length;
  let currentIndex = 0;

  while (floorIndex < ceilIndex) {
    currentIndex = Math.floor((ceilIndex - floorIndex) / 2 + floorIndex);

    if (words[currentIndex] < firstWord) {
      ceilIndex = currentIndex;
    } else {
      floorIndex = currentIndex;
    }

    if (words[currentIndex] < words[currentIndex - 1]) {
      break;
    }
  }

  return currentIndex;
};

export const solutions = {
  solution,
};

File: src/Algorithms/Lists/Rotation_Point/rotation_point.py
from math import floor


# O(log(n)) time
# O(1) space
def solution(words):
    first_word = words[0]

    floor_index = 0
    ceil_index = len(words)
    current_index = 0

    while (floor_index < ceil_index):
        current_index = floor(
            (ceil_index - floor_index) / 2 + floor_index)

        if (words[current_index] < first_word):
            ceil_index = current_index
        else:
            floor_index = current_index

        if (words[current_index] < words[current_index - 1]):
            break

    return current_index
